llama-cpp.nix

```diff
@@ -0,0 +1,118 @@
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
+
+  llama-python = python3.withPackages (ps: [ ps.numpy ps.sentencepiece ]);
+
+  # See ./overlay.nix for where these dependencies are passed in.
+  defaultBuildInputs = builtins.filter (p: p != null) [
+    Accelerate
+    MetalKit
+    CoreVideo
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
+    # HACK(Green-Sky): Nix currently has issues with CMake findcudatoolkit
+    # see https://github.com/NixOS/nixpkgs/issues/224291
+    # copied from jaxlib
+    name = "${cudatoolkit.name}-merged";
+    paths = [
+      cudatoolkit.lib
+      cudatoolkit.out
+    ] ++ optionals (versionOlder cudatoolkit.version "11") [
+      # For some reason, some of the required libs are in the targets/x86_64-linux
+      # directory; not sure why, but this works around it.
+      "${cudatoolkit}/targets/${stdenv.system}"
+    ];
+  });
+
+  rocmBuildInputs = with rocmPackages; [ clr hipblas rocblas ];
+in
+
+stdenv.mkDerivation {
+  name = "llama.cpp";
+  src = ./.;
+  meta = {
+    description = "Inference of LLaMA model in pure C/C++${descriptionSuffix}";
+    mainProgram = "llama";
+  };
+
+  postPatch = ''
+    substituteInPlace ./ggml-metal.m \
+      --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
+    substituteInPlace ./*.py --replace "/usr/bin/env python" "${llama-python}/bin/python"
+  '';
+
+  nativeBuildInputs = [ cmake ninja pkg-config ];
+
+  buildInputs = [ openmpi ]
+  ++ optional useOpenCL clblast
+  ++ optional useCuda cudaBuildInput
+  ++ optionals useRocm rocmBuildInputs
+  ++ optionals isDefault defaultBuildInputs;
+
+  cmakeFlags = [
+    "-DLLAMA_NATIVE=OFF"
+    "-DLLAMA_BUILD_SERVER=ON"
+    "-DBUILD_SHARED_LIBS=ON"
+    "-DCMAKE_SKIP_BUILD_RPATH=ON"
+  ]
+  ++ optional useOpenCL "-DLLAMA_CLBLAST=ON"
+  ++ optional useCuda "-DLLAMA_CUBLAS=ON"
+  ++ optionals useRocm [
+    "-DLLAMA_HIPBLAS=1"
+    "-DCMAKE_C_COMPILER=hipcc"
+    "-DCMAKE_CXX_COMPILER=hipcc"
+    # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
+    # in https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
+    # and select the line that matches the current nixpkgs version of rocBLAS.
+    "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
+  ]
+  ++ optionals isDefault (if (MetalKit != null) then [
+    "-DCMAKE_C_FLAGS=-D__ARM_FEATURE_DOTPROD=1"
+    "-DLLAMA_METAL=ON"
+  ] else [
+    "-DLLAMA_BLAS=ON"
+    "-DLLAMA_BLAS_VENDOR=OpenBLAS"
+  ]);
+
+  postInstall = ''
+    mv $out/bin/main $out/bin/llama
+    mv $out/bin/server $out/bin/llama-server
+    mkdir -p $out/include
+    cp $src/llama.h $out/include/
```

Comment:
It's better to add proper install targets at the cmake level, if they haven't been added yet. An example (a bit unorderly) https://github.com/SomeoneSerge/pkgs/blob/628b080165cde1528c2d40249cef114cf2cec3e2/pkgs/by-name/ll/llama-cpp/package.nix#L42-L56

Reply:
I'll leave that for another PR by someone more experienced in CMake than I. 🧑‍🎓